kernel/thread.c

```diff
@@ -36,6 +36,26 @@ extern struct _static_thread_data _static_thread_data_list_end[];
 	     thread_data < _static_thread_data_list_end; \
 	     thread_data++)
 
+#if defined(CONFIG_THREAD_MONITOR)
+void k_thread_foreach(k_thread_user_cb_t user_cb, void *user_data)
+{
+	struct k_thread *thread;
+	unsigned int key;
+
+	__ASSERT(user_cb, "user_cb can not be NULL");
+
+	thread = _kernel.threads;
+	while (thread != NULL) {
+		user_cb(thread, user_data);
+
+		key = irq_lock();
```

Comment:
This lock won't help.  It needs to surround the whole loop.

Reply:
I do not want to hold the lock while calling callback..ok will fix it.

Reply:
I know; locking interrupts for an indefinite amount of time isn't well advised.  There's not much you can do here, though.

One of the foreach functions that will leave the interrupt unlocked while calling the callback function is `net_tcp_foreach()`.  Using the same idea will not work, because it loops through an array of known size, not a linked list that can change size during iteration.

Reply:
yeah, agree. will fix it.